library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/DefaultLocationProvider.java

```diff
@@ -221,6 +221,9 @@ void onLocationFailed(@FailType int type) {
 
     @Override
     public void onLocationChanged(Location location) {
+        if (getSourceProvider().switchTaskIsRemoved() || getSourceProvider().updateRequestIsRemoved()) {
```

Comment:
I understand the reason that you return fast when update request is removed, but i don't see the point for switchTaskIsRemoved check here. 
That should be below, right before calling `getProviderSwitchTask().stop()` instead of having whole block return. Do I miss something?

Reply:
Because the `DefaultLocationProvider.onDestroy()` is executing in a whole message, which is executing by a thread's looper. And the Android's LocationManager's callback `onLocationChanged(Location location)` is also executing in a message. When these two message is queued in the same message queue of a thread's looper and the message wrapping with `DefaultLocationProvider.onDestroy()` comes in front, the blocks,  `getSourceProvider().removeSwitchTask()` and `getSourceProvider().removeUpdateRequest()`, will be executed without interruption.
 In this case, the `onLocationChanged(Location location)` is always executed after the `DefaultLocationProvider.onDestroy()`. So actually, when `getSourceProvider().switchTaskIsRemoved()` returns true, `getSourceProvider().updateRequestIsRemoved()` is always returns true here. So `switchTaskIsRemoved()` and `updateRequestIsRemoved()` are just double checking, and explicitly show the developers that `switchTaskIsRemoved` or `updateRequestIsRemoved` can both cause the callback `onLocationChanged(Location location)` failed.
But there is anther case. When developer created the Android's LocationUpdates in a looper(see [requestLocationUpdates(java.lang.String, long, float, android.location.LocationListener, android.os.Looper](https://developer.android.com/reference/android/location/LocationManager.html#requestLocationUpdates(java.lang.String, long, float, android.location.LocationListener, android.os.Looper)), and remember that, looper is always bonded with a thread) that is different from that the `DefaultLocationProvider` is in, this may cause some multi-thread synchronization problems. But your current project is not supporting this use case.

Reply:
Hey sorry for late answer, but I have to insist on moving `switchTaskIsRemoved()` check from here. Even for readability purposes. I understand your reasoning, that's true, but please move it to right before `getProviderSwitchTask().stop()`.

This is a quick fix for `onLocationChanged` since it is used the most. But the problem stands actually. I should think a more generic way to prevent this listener getting called at all. Most probably a wrapper in `DefaultLocationSource` would be necessary to achieve that so. 🤔